SQL: List data streams as VIEWs

bpintea opened the pull request:
This completes data streams support to SQL. Querying data streams has
been working transparently already, what has been missing so far is
listing them in metadata discovery.

'SHOW/SYS TABLES' will now correctly list the datastreams as VIEWs.
This will allow BI tools to keep a seamless workflow when querying
data streams.

Fixes #83449.

bpintea committed:
List data streams as VIEWs

This completes data streams support to SQL. Querying data streams has
been working transparently already, what has been missing so far is
listing them in metadata discovery.

'SHOW/SYS TABLES' will now correctly list the datastreams as VIEWs.

Commit:
Update docs/changelog/85168.yaml

Commit:
Update docs/changelog/85168.yaml

elasticsearchmachine commented:
Hi @bpintea, I've updated the changelog YAML for you.

bpintea committed:
Merge branch 'master' into enh/sql_list_ds

bpintea commented:
@elasticsearchmachine run elasticsearch-ci/part-1

bpintea committed:
Merge branch 'master' into enh/sql_list_ds

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

luigidellaquila commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/BaseRestSqlTestCase.java:
```diff
@@ -210,6 +219,24 @@ protected static void deleteIndex(String name) throws IOException {
         deleteIndex(provisioningClient(), name);
     }
 
+    protected void createDataStream(String dataStreamName) throws IOException {
```
I see this is duplicated in BaseRestSqlTestCase and JdbcIntegrationTestCase, can't we move it to a superclass?

bpintea replied:
Makes sense, I've just imported them.

luigidellaquila reviewed (commented):
LGTM, just leaving comments for minor improvements

Luegg commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java:
```diff
@@ -216,20 +217,13 @@ public void resolveNames(
+                    indexInfos.add(new IndexInfo(clusterName, alias.getName(), IndexType.ALIAS));
+                }
+                for (ResolveIndexAction.ResolvedDataStream dataStream : response.getDataStreams()) {
+                    indexInfos.add(new IndexInfo(clusterName, dataStream.getName(), IndexType.ALIAS));
```
Would it make sense to have a new `IndexType` for data streams?

bpintea replied:
It normally would, but for drivers this could be problematic with the BI tools.

costin replied:
I had initially the same comment however I agree with @bpintea - I'm not sure what value the distinction would make internally since datastreams are transparent to us. It's clear though that it would complicate slightly some code as the enum would be modified so that SQL views are mapped not just to aliases but also datastreams. Not a big change but an annoying one nevertheless.

Luegg commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/JdbcIntegrationTestCase.java:
```diff
@@ -92,6 +95,24 @@ public static void delete(String index, String documentId) throws IOException {
+        assertEquals(200, response.getStatusLine().getStatusCode());
+    }
+
+    protected static void deleteDataStream(String dataStream) throws IOException {
```
```suggestion
    protected void deleteDataStream(String dataStream) throws IOException {
```

bpintea replied:
Refactored.

bpintea committed:
Address review comments

Dedup code.

bpintea committed:
Merge branch 'master' into enh/sql_list_ds

luigidellaquila commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/JdbcIntegrationTestCase.java:
```diff
@@ -29,6 +29,8 @@
 
 public abstract class JdbcIntegrationTestCase extends RemoteClusterAwareSqlRestTestCase {
 
+    private static final String DATA_STREAM_TEMPLATE = "test-ds-index-template";
```
You don't need this anymore ;-)

Luegg reviewed (approved):
Lgtm 👍

costin reviewed (approved):
LGTM

astefan reviewed (approved):
LGTM
Left one small suggestion, but it's also ok to ignore it.

bpintea committed:
Merge branch 'master' into enh/sql_list_ds

bpintea committed:
Address review comments

- remove no longer needed declaration;
- reuse existing code for status checking.
